libafl/src/observers/stacktrace.rs

```diff
@@ -0,0 +1,319 @@
+                let map = shmem.map();
+                let mut bytes: [u8; 8] = [0; 8];
+                for i in 0..8 {
+                    bytes[i] = map[i];
+                }
+                u64::from_be_bytes(bytes)
+            }
+            Self::StaticVariable(var) => *var,
+            Self::None => panic!("BacktraceSharedMemoryWrapper is not set yet11!"),
+        }
+    }
+}
+
+// Used for fuzzers not running in the same process
+/// Static variable storing shared memory information
+pub static mut BACKTRACE_HASH_VALUE: BacktraceHashValueWrapper = BacktraceHashValueWrapper::None;
+
+/// Utilities for setting up the signal handler and panic handler to collect the backtrace
+pub mod stacktrace_hooks {
+    use crate::bolts::os::unix_signals::Signal;
+    use ahash::AHasher;
+    use backtrace::Backtrace;
+    use libc::{
+        c_int, c_void, sigaction, sigaddset, sigemptyset, siginfo_t, SA_NODEFER, SA_SIGINFO,
+        SIGALRM,
+    };
+    use std::hash::Hasher;
+    use std::{mem, panic, ptr};
+
+    /// Collects the backtrace via Backtrace and Debug
+    /// Debug used for dev purposes, will hash symbols later
+    pub fn collect_backtrace() {
+        let b = Backtrace::new();
+        // will use symbols later
+        let trace = format!("{:?}", b);
+        eprintln!("{}", trace);
+        let mut hasher = AHasher::new_with_keys(0, 0);
+        hasher.write(trace.as_bytes());
+        let hash = hasher.finish();
+        println!("backtrace collected with hash={}", hash);
+        unsafe {
+            crate::observers::BACKTRACE_HASH_VALUE.store_stacktrace_hash(hash);
+        }
+    }
+
+    /// setup backtrace collection in a rust panic hook when the harness is rust code
+    pub fn setup_rust_panic_hook() {
+        panic::set_hook(Box::new(|_panic_info| {
+            collect_backtrace();
+        }));
+    }
+
+    /// setup backtrace collection in a signal handler when the harness is linked via FFI
+    pub unsafe fn setup_signal_handler() {
+        println!("setting up stacktrace signal handler");
+        fn signal_handler(sig: c_int, _info: siginfo_t, _con: *mut c_void) {
+            println!("Received signal sig={}", sig);
+            collect_backtrace();
+        }
+        let signals = vec![
+            Signal::SigAlarm,
+            Signal::SigUser2,
+            Signal::SigAbort,
+            Signal::SigBus,
+            Signal::SigPipe,
+            Signal::SigFloatingPointException,
+            Signal::SigIllegalInstruction,
+            Signal::SigSegmentationFault,
+            Signal::SigTrap,
+        ];
+        let mut sa: sigaction = mem::zeroed();
+        sigemptyset(&mut sa.sa_mask as *mut libc::sigset_t);
+        sigaddset(&mut sa.sa_mask as *mut libc::sigset_t, SIGALRM);
+        sa.sa_sigaction = signal_handler as usize;
+        sa.sa_flags = SA_NODEFER | SA_SIGINFO;
+        for sig in signals {
+            sigaction(sig as i32, &mut sa as *mut sigaction, ptr::null_mut());
```

Comment:
This conflicts with InProcessExecutor that install a signal handler (btw, we have rust-y API for that in bolts) to recover on crash. You can simply call `collect_backtrace();` in the observer post_exec. Yes, you will trace also post_exec, but it's not an issue and if you want you can remove the first element of the trace after collecting

Reply:
Yes right, but we can add a parameter to post_exec, the exit_kind. In this way, you can collect on `if exit_kind == ExitKind::Crash`

Reply:
Done